qutip/core/cy/_element.pyx

```diff
@@ -356,28 +357,67 @@ cdef class _EvoElement(_BaseElement):
 
 cdef class _FuncElement(_BaseElement):
     """
-    Used with :obj:`QobjEvo` to build a function with signature: ::
+    Used with :obj:`QobjEvo` to build an evolution term from a function with
+    either the signature: ::
 
-      func(t: float, args: dict) -> Qobj
+        func(t: float, ...) -> Qobj
 
-    :obj:`QobjEvo` created from such a function contian one
-    :obj:`_FuncElement`::
+    or the older QuTiP 4 style signature: ::
 
-      qevo = QobjEvo(func, args=args)
-      qevo.elements = [_FuncElement(func, args)]
+        func(t: float, args: dict) -> Qobj
 
-    Each :obj:`_FuncElement` contain an immutable pair ``(func, args)``.
+    In the new style, ``func`` may accept arbitrary named arguments and
+    is called as ``func(t, **args)``.
 
-    This class has basic memoize capacity: it saves the last call ::
+    A :obj:`QobjEvo` created from such a function contains one
+    :obj:`_FuncElement`: ::
+
+        qevo = QobjEvo(func, args=args)
+        qevo.elements = [_FuncElement(func, args)]
+
+    Each :obj:`_FuncElement` consists of an immutable pair ``(func, args)``
+    of function and argument.
+
+    This class has a basic capacity to memoize calls to ``func`` by saving the
+    result of the last call ::
 
         op = QobjEvo(func, args=args)
         (op.dag() * op)(t)
 
-    call the function `func` only once.
+    calls ``func`` only once.
+
+    Parameters
+    ----------
+    func : callable(t : float, ...) -> Qobj
+        Function returning the element value at time ``t``.
+
+    args : dict
+        Values of the arguments to pass to ``func``.
+
+    style : {None, "pythonic", "dict", "auto"}
+        The style of the signature used. If style is ``None``,
+        the value of ``qutip.settings.core["function_coefficient_style"]``
+        is used. Otherwise the supplied value overrides the global setting.
+
+    The parameters ``_f_pythonic`` and ``_f_parameters`` override function
+    style and parameter detection and are not intended to be part of
+    the public interface.
     """
-    def __init__(self, func, args):
+
+    _UNSET = object()
+
+    def __init__(self, func, args, style=None, bint _f_pythonic=_UNSET, _f_parameters=_UNSET):
+        if _f_pythonic is self._UNSET or _f_parameters is self._UNSET:
+            _f_pythonic, _f_parameters = coefficient_function_parameters(
+                func, style=style)
```

Comment:
I added `_UNSET` because `_f_parameters=None` has the special meaning "all parameters accepted".

I used `_UNSET` for `_f_pythonic` too for consistency with `_f_parameters` because they are used together.

I forgot to remove the `bint` in the function signature here -- thank you for catching that.

I think the extra type check for when only one is _UNSET is a good idea, but I will rework it a little so that only one check happens on the fast-path through the constructor.

Reply:
Implemented.